lighthouse/aggregate-service-import.md

```diff
@@ -0,0 +1,137 @@
+  pod IP addresses from the originating cluster.
+
+## Proposal
+
+Submariner has a passive broker, it is a central apiserver that is accessed by each cluster in
+the clusterset. Clusters synchronise resources to and from the broker with all sync and
+reconciliation logic implemented in the clusters.
+
+In this proposal we retain the passive broker and take an approach to merging `ServiceImports`
+where exporting clusters handle the merging. Simplistically, a merged `ServiceImport` outcome
+can be achieved by allowing the first exporting cluster to create a `ServiceImport` and all
+subsequent exporting clusters to reuse it, assuming they have matching port lists.
+
+1. First exporting cluster creates a `ServiceImport` on the broker.
+1. Subsequent exporting clusters reuse the existing `ServiceImport`.
+1. Exporting clusters create and sync `EndpointSlices` to the broker with the service VIP and a
+   `multicluster.kubernetes.io/service-name` label that matches the `ServiceImport`.
+1. All clusters sync the `ServiceImport` and `EndpointSlices` from the broker.
+
+Looking at the `ServiceImport` merging in more detail, the MCS specification asks for all port
+definitions from the `ServiceExports` to be combined in the merged `ServiceImport`. This means
+that an exporting cluster would need to merge its port definitions into the `ServiceImport`.
+Conversely, when removing a `ServiceExport` any merged port definitions would need to be
+removed. However an exporting cluster only has visibility of its local `ServiceExport` so cannot
+unmerge with the information available locally.
+
+Note: It is an open question what behaviour is expected when a connection request for service
+port is directed to a cluster that does not expose that port.
+
+The notion that services are equivalent when they have matching name and namespace but different
+port lists seems problematic. It is preferable to set `ServiceExport` status to conflict for any
+differences between the exported services. By choosing to be strict about service equivalence,
+we avoid having to perform any port list merging or unmerging on the `ServiceImport`.
+
+### DNS Resolution
+
+Lighthouse currently implements DNS resolution by round-robin load balancing between the
+`ServiceImports` that match the namespace/name pair. With a merged `ServiceImport`, the
+per-cluster service VIPs are stored on the `EndpointSlices`. DNS resolution will select from the
+`EndpointSlices` associated with the `ServiceImport` instead of using `ServiceImports`
+directly.
+
+DNS Resolution behaviour for headless services will remain unchanged.
+
+### Service VIPs in Importing Clusters
+
+Given the way that Submariner operates, there is no need to assign a Service VIP to a
+`ServiceImport` in the importing clusters. This means that `ServiceImports` can always have an
+empty IP address attribute, like they do for headless services today.
```

Comment:
We can think of eventually adding VIP as well so that an it can work any component that consumes MCS API. May be globalnet can help in providing an IP for this.

Reply:
I think the goal would be to perform load balancing across the workload clusters behind a cluster local Service VIP, so this would be a move from DNS load balancing to a ServiceImport load balancer. I suggest that it be tracked as follow-on work that is out of scope for this enhancement.

Reply:
Yes, wouldbe good to have this option but it will have to be a config option that can be enabled/diabled. At least for now we don't have any plans to implement an mcs controller that will loadbalance using endpointslices in datapath instead of DNS, but it will allow us to use Lighthouse-only deployments to provide discovery for other cluster connectivity solutions.